Initial offline mode ADR

cardenaso11 opened the pull request:
Making this draft publicly visible before the full PR etc, so everyone can see what to prepare for ahead of time. The current status of the implementation, is that it is nearing completion.

This ADR is for support for an offline mode, for the hydra-node target. SundaeSwap intends to contribute these patches upstream. Feedback is welcome, if there are any questions or suggestions or problems.

---

<!-- Consider each and tick it off one way or the other -->
* [ ] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [ ] Haddocks updated or not needed
* [ ] No new TODOs introduced or explained hereafter

cardenaso11 committed:
initial offline mode adr draft

cardenaso11 commented:
> Some comments left, I think we could probably pair or work in team on this ADR to speed up the feedback loop

Thank you! Taking a look now. Making some initial adjustments based on feedback, will message about pairing. Sounds good to me

ghost commented:
Following our discussion, an updated lightweight architecture diagram taking into account what would be dropped/adapted: 

![ADRs _ Designs - Offline mode ADR](https://github.com/input-output-hk/hydra/assets/79840582/caaffdaf-0564-437a-8845-7ac8f62d9c04)

cardenaso11 committed:
revise offline ADR a bit and add very tenative persistence adr sketch

ghost commented:
@cardenaso11 Do you want more feedback on this? Is it ready for review, or do you want to make the ADR part of the PR introducing the feature?

cardenaso11 committed:
clarify actions

cardenaso11 committed:
wip: explain dependency injection, decide if needs another adr

ghost commented:
Follow-up discussion outcome:
* Implement a "mock chain" by providing a `Chain m tx` handler and do not bother with mocking actual Cardano interactions
* "Dependency injection" Is not really needed in `Direct` but could improve readability and maintainability of `hydra-cluster` code
* Old mock-chain code from the early days of Hydra could be useful inspiration https://github.com/input-output-hk/hydra/blob/00026ec59f4f6f6d7a535816edd003ef9b81d04e/local-cluster/src/HydraNode.hs#L207
* Polishing the off-line ADR
* Need more work to refine the need for the event streaming ADR

cardenaso11 committed:
discussion with hydra folks

cardenaso11 committed:
remove redundant comments

cardenaso11 committed:
remove additional adr from this PR so this can be reviewed standalone

ghost commented:
@cardenaso11 I would like to move this forward, I think the off-line ADR is in good shape and it's in `Proposed` state anyway, until the PR comes in. Can you remove the other one?

cardenaso11 commented:
> @cardenaso11 I would like to move this forward, I think the off-line ADR is in good shape and it's in `Proposed` state anyway, until the PR comes in. Can you remove the other one? 

Absolutely, thanks a ton. I'd done that locally but think I had not pushed it

cardenaso11 commented:
GitHub UI somehow rebased like 3 times instead of one for some reason, but either way, removed the other ADR. I also added the edits based on our meeting, let me know if anything else would help. Thanks a ton, @abailly-iohk

cardenaso11 commented on docs/adr/2023-07-05_024-offline-mode.md:
```diff
@@ -0,0 +1,32 @@
+
+Commit endpoint will return 400 instead of building a transaction, in offline mode.
+
+<!-- NOTE: do we want to support having peer hydra nodes in offline/mock chain mode? Would need to add networking to replace communication done over L1 -->
```
Agree on this. After thinking about it for a bit, it's something that could be added later if there's a usecase for it, but I think currently it doesn't seem important enough to block on it

v0d1ch commented on docs/adr/2023-07-05_024-offline-mode.md:
```diff
@@ -0,0 +1,32 @@
+## Decision
+Hydra node will be executable in offline mode, as an alternative to the default online mode. When online, the Hydra node depends on querying a Cardano node for Era History information and Genesis parameters. When offline this is not necessary, because the Hydra node will not connect to any Layer 1.
+
+The initial state of the head will be specified in a flag, which makes any Commit redundant. The flag will specify a file for the starting Layer 2 UTXO. The Hydra node can be configured to wrte the current UTXO into a file, including the starting UTXO file.
```
```suggestion
The initial state of the head will be specified in a flag, which makes any Commit redundant. The flag will specify a file for the starting Layer 2 UTXO. The Hydra node can be configured to write the current UTXO into a file, including the starting UTXO file.
```

v0d1ch reviewed (commented):
@cardenaso11 This looks good to me. Not sure if we want to specify exactly how to get to the open Head with this initial utxo we will have as the argument to hydra-node and bypass `Commit`. Perhaps this would be too much information for an adr and can be left for the implementation.

cardenaso11 committed:
address current feedback, sans gummiworm docs temporarily

cardenaso11 committed:
gummiworm explanation approved by management, typo fixes

ghost commented:
@cardenaso11 Apart from some minor comments, we are good to go :)

cardenaso11 committed:
adr number bump

cardenaso11 commented:
Great! Let me know if there's anything else remaining on this ADR to get merged, in the meantime I am gonna make sure the actual cleaned up implementation branch is reviewable for you guys by Monday. Thanks a ton!

ghost commented:
@cardenaso11 Seems like you need to add your handle to the `authors.yml` :)

cardenaso11 committed:
Update authors.yaml

add cardenaso11 (self) to authors.yaml
